src/model.rs

```diff
@@ -206,62 +223,218 @@ impl Analysis<Mdl> for TensorAnalysis {
                 let t_a = x(a).meta;
 
                 unsafe {
-                    let relu = g.relu(t_a, true);
-                    //let r_cost = (*(*relu).op.ptr).runtime;
-                    //println!("Cost of relu is {}", r_cost);
+                    let res = g.relu(t_a, true);
+                    Self::Data {
+                        dtype: DataKind::Tnsr,
+                        val: 0,
+                        name: String::new(),
+                        meta: res,
+                    }
+                }
+            }
+
+            Mdl::Tanh(a) => {
+                assert!(x(a).dtype == DataKind::Tnsr);
+                let t_a = x(a).meta;
+
+                unsafe {
+                    let res = g.tanh(t_a, true);
                     Self::Data {
                         dtype: DataKind::Tnsr,
                         val: 0,
-                        meta: relu,
+                        name: String::new(),
+                        meta: res,
                     }
                 }
             }
 
-            Mdl::Input([name, dim1, dim2, dim3, dim4]) => {
+            Mdl::Sigmoid(a) => {
+                assert!(x(a).dtype == DataKind::Tnsr);
+                let t_a = x(a).meta;
+
+                unsafe {
+                    let res = g.sigmoid(t_a, true);
+                    Self::Data {
+                        dtype: DataKind::Tnsr,
+                        val: 0,
+                        name: String::new(),
+                        meta: res,
+                    }
+                }
+            }
+
+            Mdl::Input([name]) => {
+                // Check types
                 assert!(x(name).dtype == DataKind::Name);
-                assert!(x(dim1).dtype == DataKind::Scalar);
-                assert!(x(dim2).dtype == DataKind::Scalar);
-                assert!(x(dim3).dtype == DataKind::Scalar);
-                assert!(x(dim4).dtype == DataKind::Scalar);
 
+                // Get shape
+                let mut split = x(name).name.split("@");
```

Comment:
Why `mut`? You can write this as `let split = x(name).name.split("@").collect()`. Idem for `dim_str_vec`.

Reply:
Updated